File: apps/client/lib/data/api/sdkproto/v1/permissions.dart
// PermissionUpdate 6 union variant + PermissionResult。

import 'package:json_annotation/json_annotation.dart';

part 'permissions.g.dart';

class PermissionBehavior {
  static const allow = 'allow';
  static const deny = 'deny';
  static const ask = 'ask';
}

class PermissionMode {
  static const defaultMode = 'default';
  static const acceptEdits = 'acceptEdits';
  static const bypassPermissions = 'bypassPermissions';
  static const plan = 'plan';
  static const dontAsk = 'dontAsk';
}

class PermissionDestination {
  static const userSettings = 'userSettings';
  static const projectSettings = 'projectSettings';
  static const localSettings = 'localSettings';
  static const session = 'session';
  static const cliArg = 'cliArg';
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class PermissionRuleValue {
  final String toolName;
  final String? ruleContent;
  PermissionRuleValue({required this.toolName, this.ruleContent});
  factory PermissionRuleValue.fromJson(Map<String, dynamic> json) =>
      _$PermissionRuleValueFromJson(json);
  Map<String, dynamic> toJson() => _$PermissionRuleValueToJson(this);
}

abstract class PermissionUpdate {
  String get type;
  Map<String, dynamic> toJson();

  static PermissionUpdate fromJson(Map<String, dynamic> json) {
    final t = json['type'] as String? ?? '';
    switch (t) {
      case 'addRules':
        return AddRules.fromJson(json);
      case 'replaceRules':
        return ReplaceRules.fromJson(json);
      case 'removeRules':
        return RemoveRules.fromJson(json);
      case 'setMode':
        return SetModeUpdate.fromJson(json);
      case 'addDirectories':
        return AddDirectories.fromJson(json);
      case 'removeDirectories':
        return RemoveDirectories.fromJson(json);
      default:
        throw ArgumentError('unknown PermissionUpdate type: $t');
    }
  }
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class AddRules implements PermissionUpdate {
  @override
  final String type;
  final List<PermissionRuleValue> rules;
  final String behavior;
  final String destination;

  AddRules({
    this.type = 'addRules',
    required this.rules,
    required this.behavior,
    required this.destination,
  });

  factory AddRules.fromJson(Map<String, dynamic> json) =>
      _$AddRulesFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$AddRulesToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class ReplaceRules implements PermissionUpdate {
  @override
  final String type;
  final List<PermissionRuleValue> rules;
  final String behavior;
  final String destination;

  ReplaceRules({
    this.type = 'replaceRules',
    required this.rules,
    required this.behavior,
    required this.destination,
  });

  factory ReplaceRules.fromJson(Map<String, dynamic> json) =>
      _$ReplaceRulesFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$ReplaceRulesToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class RemoveRules implements PermissionUpdate {
  @override
  final String type;
  final List<PermissionRuleValue> rules;
  final String behavior;
  final String destination;

  RemoveRules({
    this.type = 'removeRules',
    required this.rules,
    required this.behavior,
    required this.destination,
  });

  factory RemoveRules.fromJson(Map<String, dynamic> json) =>
      _$RemoveRulesFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$RemoveRulesToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class SetModeUpdate implements PermissionUpdate {
  @override
  final String type;
  final String mode;
  final String destination;

  SetModeUpdate({
    this.type = 'setMode',
    required this.mode,
    required this.destination,
  });

  factory SetModeUpdate.fromJson(Map<String, dynamic> json) =>
      _$SetModeUpdateFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$SetModeUpdateToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class AddDirectories implements PermissionUpdate {
  @override
  final String type;
  final List<String> directories;
  final String destination;

  AddDirectories({
    this.type = 'addDirectories',
    required this.directories,
    required this.destination,
  });

  factory AddDirectories.fromJson(Map<String, dynamic> json) =>
      _$AddDirectoriesFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$AddDirectoriesToJson(this);
}

@JsonSerializable(includeIfNull: false, anyMap: true)
class RemoveDirectories implements PermissionUpdate {
  @override
  final String type;
  final List<String> directories;
  final String destination;

  RemoveDirectories({
    this.type = 'removeDirectories',
    required this.directories,
    required this.destination,
  });

  factory RemoveDirectories.fromJson(Map<String, dynamic> json) =>
      _$RemoveDirectoriesFromJson(json);
  @override
  Map<String, dynamic> toJson() => _$RemoveDirectoriesToJson(this);
}

/// PermissionResult —— allow / deny 两态合并到单 class。
@JsonSerializable(includeIfNull: false, anyMap: true)
class PermissionResult {
  final String behavior;
  final dynamic updatedInput;
  final List<dynamic>? updatedPermissions;
  final String? message;
  final bool? interrupt;

  PermissionResult({
    required this.behavior,
    this.updatedInput,
    this.updatedPermissions,
    this.message,
    this.interrupt,
  });

  factory PermissionResult.fromJson(Map<String, dynamic> json) =>
      _$PermissionResultFromJson(json);
  Map<String, dynamic> toJson() => _$PermissionResultToJson(this);

  bool get isAllow => behavior == PermissionBehavior.allow;
  bool get isDeny => behavior == PermissionBehavior.deny;
}

File: packages/go-sdk/biu/sdkproto/v1/permissions.go
package sdkproto

import (
	"encoding/json"
	"fmt"
)

// Permission enum 字面量。
const (
	PermissionAllow = "allow"
	PermissionDeny  = "deny"
	PermissionAsk   = "ask"

	PermissionModeDefault           = "default"
	PermissionModeAcceptEdits       = "acceptEdits"
	PermissionModeBypassPermissions = "bypassPermissions"
	PermissionModePlan              = "plan"
	PermissionModeDontAsk           = "dontAsk"

	PermissionDestUserSettings    = "userSettings"
	PermissionDestProjectSettings = "projectSettings"
	PermissionDestLocalSettings   = "localSettings"
	PermissionDestSession         = "session"
	PermissionDestCliArg          = "cliArg"

	PermissionUpdateAddRules          = "addRules"
	PermissionUpdateReplaceRules      = "replaceRules"
	PermissionUpdateRemoveRules       = "removeRules"
	PermissionUpdateSetMode           = "setMode"
	PermissionUpdateAddDirectories    = "addDirectories"
	PermissionUpdateRemoveDirectories = "removeDirectories"
)

type PermissionRuleValue struct {
	ToolName    string `json:"toolName"`
	RuleContent string `json:"ruleContent,omitempty"`
}

// PermissionUpdate 是 6 个变体的标记接口。type 字段做 discriminator。
type PermissionUpdate interface {
	isPermissionUpdate()
	PermissionUpdateType() string
}

type AddRules struct {
	Type        string                `json:"type"` // "addRules"
	Rules       []PermissionRuleValue `json:"rules"`
	Behavior    string                `json:"behavior"`
	Destination string                `json:"destination"`
}

func (*AddRules) isPermissionUpdate()          {}
func (*AddRules) PermissionUpdateType() string { return PermissionUpdateAddRules }

type ReplaceRules struct {
	Type        string                `json:"type"` // "replaceRules"
	Rules       []PermissionRuleValue `json:"rules"`
	Behavior    string                `json:"behavior"`
	Destination string                `json:"destination"`
}

func (*ReplaceRules) isPermissionUpdate()          {}
func (*ReplaceRules) PermissionUpdateType() string { return PermissionUpdateReplaceRules }

type RemoveRules struct {
	Type        string                `json:"type"` // "removeRules"
	Rules       []PermissionRuleValue `json:"rules"`
	Behavior    string                `json:"behavior"`
	Destination string                `json:"destination"`
}

func (*RemoveRules) isPermissionUpdate()          {}
func (*RemoveRules) PermissionUpdateType() string { return PermissionUpdateRemoveRules }

type SetModeUpdate struct {
	Type        string `json:"type"` // "setMode"
	Mode        string `json:"mode"`
	Destination string `json:"destination"`
}

func (*SetModeUpdate) isPermissionUpdate()          {}
func (*SetModeUpdate) PermissionUpdateType() string { return PermissionUpdateSetMode }

type AddDirectories struct {
	Type        string   `json:"type"` // "addDirectories"
	Directories []string `json:"directories"`
	Destination string   `json:"destination"`
}

func (*AddDirectories) isPermissionUpdate()          {}
func (*AddDirectories) PermissionUpdateType() string { return PermissionUpdateAddDirectories }

type RemoveDirectories struct {
	Type        string   `json:"type"` // "removeDirectories"
	Directories []string `json:"directories"`
	Destination string   `json:"destination"`
}

func (*RemoveDirectories) isPermissionUpdate()          {}
func (*RemoveDirectories) PermissionUpdateType() string { return PermissionUpdateRemoveDirectories }

// UnmarshalPermissionUpdate dispatch 6 个变体。
func UnmarshalPermissionUpdate(data []byte) (PermissionUpdate, error) {
	var head struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(data, &head); err != nil {
		return nil, fmt.Errorf("permission_update: peek type: %w", err)
	}
	var v PermissionUpdate
	switch head.Type {
	case PermissionUpdateAddRules:
		v = &AddRules{}
	case PermissionUpdateReplaceRules:
		v = &ReplaceRules{}
	case PermissionUpdateRemoveRules:
		v = &RemoveRules{}
	case PermissionUpdateSetMode:
		v = &SetModeUpdate{}
	case PermissionUpdateAddDirectories:
		v = &AddDirectories{}
	case PermissionUpdateRemoveDirectories:
		v = &RemoveDirectories{}
	default:
		return nil, fmt.Errorf("permission_update: unknown type %q", head.Type)
	}
	if err := json.Unmarshal(data, v); err != nil {
		return nil, fmt.Errorf("permission_update: unmarshal %s: %w", head.Type, err)
	}
	return v, nil
}

// PermissionResult 是 can_use_tool 的响应：allow 或 deny。
// behavior 区分；allow 带 updatedInput / updatedPermissions，deny 带 message / interrupt。
// 用单 struct 表示两态，简化 marshal。
type PermissionResult struct {
	Behavior           string             `json:"behavior"`
	UpdatedInput       json.RawMessage    `json:"updatedInput,omitempty"`
	UpdatedPermissions []PermissionUpdate `json:"updatedPermissions,omitempty"`
	Message            string             `json:"message,omitempty"`
	Interrupt          *bool              `json:"interrupt,omitempty"`
}
